### PR TITLE
Add GitHub Workflow action for building and testing

### PR DESCRIPTION
**Pull request by themakshter**
This change re-creates the build and test steps currently being run in travis

### CODE REVIEW

**Comment by themakshter** [comment]
I'm working on the draft, but I need access to some of the information to be able to test this locally in my own repository before I'm certain it will work correctly.

Namely:
- read access on the `branding` repo
- ability to deploy

Of course, deploying to the live site would be horrible - could we set up a similar structure (exactly the same but at a different address etc.) whose environment variables we could inject to test the deploy?

**Comment by drimpact** [comment]
I have given you read access to the branding repo.
The following env vars has been configured in actions: AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY and CLOUDFRONT_DISTRIBUTION_ID.
These credentials have permission to write to the `impact-web-dev` bucket in S3, this bucket is public, so you can verify the writes. I have not replicated cloudfront, but you are welcome to do invalidations on the main cloudfront - there is no impact - and I can confirm they are happening.

**Comment by drimpact** [comment]
I just set AWS_S3_BUCKET too, so that we don't have to hard code the bucket name

File: config/test/karma.conf.js [hunk 1]
```diff
@@ -35,7 +35,7 @@ module.exports = function (config) {
 
     autoWatch: true,
 
-    singleRun: false,
+    singleRun: true,
```
Comment (themakshter) [inline config/test/karma.conf.js:38]: If I don't do this, the test hangs

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v2
+      - name: Remove previous branding folder
+        run: rm -rf ./src/branding
+      - name: Checkout branding repository
+        uses: actions/checkout@v2
+        with:
+          repository: impactasaurus/branding
+          token: ${{ secrets.BRANDING_PERSONAL_ACCESS_TOKEN }}
```
Comment (themakshter) [inline .github/workflows/ci.yaml:23]: We need this Personal access token for access the private repo: https://github.com/actions/checkout#checkout-multiple-repos-private

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        uses: actions/setup-node@v1
+        with:
+          node-version: 11.10.0
+      - name: Install dependencies
+        run: npm ci
+      - name: Run Linter
+        run: npm run lint
+      - name: Run Tests
+        run: npm test
+      - name: Generate Production Files
```
Comment (themakshter) [inline .github/workflows/ci.yaml:35]: Would you rather create normal debug files in PRs?

**Review by themakshter — COMMENTED** [review]
Added some questions/comments.

I have another idea to split up jobs, will demonstrate in another change

**Comment by themakshter** [comment]
Successful run: https://github.com/themakshter/app/actions/runs/522079285

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (themakshter) [reply .github/workflows/ci.yaml:23]: Could you add this as well and name it `BRANDING_PERSONAL_ACCESS_TOKEN `

Comment (drimpact) [reply .github/workflows/ci.yaml:23]: Added, sorry for the delay

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        run: npm run build:prod
+      - name: Upload build artifacts
+        uses: actions/upload-artifact@v2
+        with:
+          name: app
+          path: build
+  deploy:
+    name: Deploy App
+    if: github.ref == 'refs/heads/master'
```
Comment (themakshter) [inline .github/workflows/ci.yaml:45]: Conditional to only run when it's master branch

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+      - name: Download build artifacts
+        uses: actions/download-artifact@v2
+        with:
+          name: app
+      - name: Deploy app to S3
+        uses: reggionick/s3-deploy@v3
+          with:
+            folder: app
+            bucket: ${{ secrets.S3_BUCKET }}
+            bucket-region: ${{ secrets.S3_BUCKET_REGION }}
```
Comment (themakshter) [inline .github/workflows/ci.yaml]: Needs bucket region defined as well

File: .github/workflows/ci.yaml [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    env:
+      AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
+      AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+    steps:
+      - name: Download build artifacts
+        uses: actions/download-artifact@v2
+        with:
+          name: app
+      - name: Deploy app to S3
+        uses: reggionick/s3-deploy@v3
```
Comment (themakshter) [inline .github/workflows/ci.yaml:57]: https://github.com/Reggionick/s3-deploy I found exactly what you want haha

**Review by themakshter — COMMENTED** [review]
Added some comments for change.

Test for deploy workflow skipped when it is a PR branch: https://github.com/themakshter/app/pull/2

**Comment by themakshter** [comment]
@drimpact it's ready for review - although I haven't been able to test the deploy functionality on my side

File: .github/workflows/ci.yaml (hunk 5 above)
Comment (drimpact) [reply .github/workflows/ci.yaml]: This has been added

File: .github/workflows/ci.yaml [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        with:
+          folder: app
+          bucket: ${{ secrets.S3_BUCKET }}
+          bucket-region: ${{ secrets.S3_BUCKET_REGION }}
+          dist-id: ${{ secrets.CLOUDFRONT_DISTRIBUTION_ID }}
+          invalidation: /
+          delete-removed: true
+          no-cache: true
```
Comment (drimpact) [inline .github/workflows/ci.yaml:65]: Looking at the docs, I think this should be false, I want users to be able to cache the files

**Comment by drimpact** [comment]
It is looking really neat. Just left one comment.
I assume you can't test it because it is in another repo and hence secrets won't be shared. If that is the case, I figure we merge to master with the dev variables setup, test it works, then remove travis config, then change variables. Sound like a plan?

**Comment by themakshter** [comment]
> It is looking really neat. Just left one comment.
> I assume you can't test it because it is in another repo and hence secrets won't be shared. If that is the case, I figure we merge to master with the dev variables setup, test it works, then remove travis config, then change variables. Sound like a plan?

Yeah, the alternative would be for you to add them to my repo but that would probably not be a good idea 😄

**Pull request merged**